src/graph/sampler.cc

```diff
@@ -988,415 +988,769 @@ std::vector<dgl_id_t> Global2Local(const std::vector<size_t> &ids,
-    unchanged = dst_data;
-    neg_unchanged = neg_dst_data;
-    neg_changed = neg_src_data;
-  } else {
-    // corrupt tail nodes.
-    unchanged = src_data;
-    neg_unchanged = neg_src_data;
-    neg_changed = neg_dst_data;
+class UniformEdgeSamplerObject: public EdgeSamplerObject {
+public:
+  explicit UniformEdgeSamplerObject(const GraphPtr gptr,
+                                    IdArray seed_edges,
+                                    const int64_t batch_size,
+                                    const int64_t num_workers,
+                                    const std::string neg_mode,
+                                    const int64_t neg_sample_size,
+                                    const bool exclude_positive,
+                                    const bool check_false_neg,
+                                    IdArray relations)
+                                    : EdgeSamplerObject{gptr,
+                                        seed_edges,
+                                        batch_size,
+                                        num_workers,
+                                        neg_mode,
+                                        neg_sample_size,
+                                        exclude_positive,
+                                        check_false_neg,
+                                        relations} {
+    batch_curr_id_ = 0;
+    num_seeds_ = seed_edges->shape[0];
+    max_batch_id_ = (num_seeds_ + batch_size - 1) / batch_size;
   }
+  ~UniformEdgeSamplerObject() {}
 
-  // We first sample all negative edges.
-  std::vector<size_t> neg_vids;
-  RandomSample(num_tot_nodes,
-               num_chunks * neg_sample_size,
-               &neg_vids);
-
-  dgl_id_t curr_eid = 0;
-  std::unordered_map<dgl_id_t, dgl_id_t> neg_map;
-  dgl_id_t local_vid = 0;
-  // Collect nodes in the positive side.
-  std::vector<dgl_id_t> local_pos_vids;
-  local_pos_vids.reserve(num_pos_edges);
-  for (int64_t i = 0; i < num_pos_edges; i++) {
-    dgl_id_t vid = induced_vid_data[unchanged[i]];
-    auto it = neg_map.find(vid);
-    if (it == neg_map.end()) {
-      local_pos_vids.push_back(local_vid);
-      neg_map.insert(std::pair<dgl_id_t, dgl_id_t>(vid, local_vid++));
-    }
-  }
+  void fetch(DGLRetValue* rv) {
+    const int64_t num_workers = std::min(num_workers_, max_batch_id_ - batch_curr_id_);
+    // generate subgraphs.
+    std::vector<SubgraphRef> positive_subgs(num_workers);
+    std::vector<SubgraphRef> negative_subgs(num_workers);
+#pragma omp parallel for
+    for (int64_t i = 0; i < num_workers; i++) {
+      const int64_t start = (batch_curr_id_ + i) * batch_size_;
+      const int64_t end = std::min(start + batch_size_, num_seeds_);
+      const int64_t num_edges = end - start;
+      IdArray worker_seeds = seed_edges_.CreateView({num_edges}, DLDataType{kDLInt, 64, 1},
+                                                   sizeof(dgl_id_t) * start);
+      EdgeArray arr = gptr_->FindEdges(worker_seeds);
```

Comment:
can you make sure the ImmutableGraph has coo format before entering the openmp for loop? you can simply call GetCOO() before the for loop.

Reply:
ImmutableGraph->FindEdges() will automatically do GetCOO()->FindEdges(eids);

Reply:
I know. But when you call GetCOO() inside the openmp for loop and if the coo graph doesn't exist, every thread tries to generate the coo graph and assign it to the coo pointer. it may lead to race condition. i got this bug before. When I wrote this sampler, I forgot to do so.

Reply:
I see

Reply:
I don’t know how you like to refactor it. Currently, we dynamic cast the pointer to ImmutableGraph.